examples-kotlin/src/main/kotlin/org/bitcoinj/ForwardingService.kt

```diff
@@ -114,72 +123,55 @@ class ForwardingService(val config: Config, val wallet: Wallet) : Closeable {
         return wallet.sendTransaction(sendRequest)
             .thenCompose { broadcast: TransactionBroadcast ->
                 println("Transaction ${broadcast.transaction().txId} is signed and is being delivered to ${wallet.network()}...\n")
                 broadcast.awaitRelayed() // Wait until peers report they have seen the transaction
             }
+            .whenComplete { broadcast: TransactionBroadcast?, throwable: Throwable? ->
+                if (broadcast != null) {
+                    println("Sent ${broadcast.transaction().outputSum.toFriendlyString()} onwards and acknowledged by peers, via transaction ${broadcast.transaction().txId}\n")
+                }
+                if (throwable != null) {
+                    println("Exception occurred: $throwable")
+                }
+            }
     }
 
-    /**
-     * Create a CoinSelector that only returns outputs from a given parent transaction.
-     *
-     *
-     * This is using the idea of partial function application to create a 2-argument function for coin selection
-     * with a third, fixed argument of the transaction id.
-     * @param parentTxId The parent transaction hash
-     * @return a coin selector
-     */
-    fun forwardingCoinSelector(parentTxId: Sha256Hash): CoinSelector {
-        requireNonNull(parentTxId)
-        return CoinSelector.fromPredicate { output: TransactionOutput ->
-            output.parentTransactionHash == parentTxId
-        }
+    fun status(): String? {
+        return String.format("Waiting to receive coins on: %s", wallet.currentReceiveAddress())
     }
 
     /**
      * Close the service.
      */
     override fun close() {
         wallet.removeCoinsReceivedEventListener(::coinForwardingListener)
+        walletAppKit.close()
     }
 }
 
-fun getPrefix(network: BitcoinNetwork?): String {
-    return String.format("forwarding-service-%s", network.toString())
-}
-
-fun parseArgs(args: Array<String>): Array<Any> {
-    val USAGE = "Usage: address-to-forward-to [mainnet|testnet|signet|regtest]"
-    if (args.size < 1 || args.size > 2) {
-        System.err.println(USAGE)
-        System.exit(1)
+data class Config(
+    val forwardingAddress: Address,                                              // Address to forward to
+    val network: BitcoinNetwork = forwardingAddress.network() as BitcoinNetwork, // Network to operate on
+    val walletDirectory: File = File("."),                              // Directory to create wallet files in
+    val walletPrefix: String = getPrefix(network),                               // Prefix for wallet file names
+    val requiredConfirmations: Int = 1,                                          // Required number of tx confirmations before forwarding
+    val maxConnections: Int = 4                                                  // Maximum number of connections
+) {
+    init {
+        network.checkAddress(forwardingAddress)
```

Comment:
It's nice how the Kotlin `data class` can be built with default values and not require extra constructors, but doing this validation here seems a little weird, for two reasons:
1) It seems like the check is run after the member fields are assigned.
2) It appears to run `checkAddress` redundantly in the case where `network` is read from `forwardingAddress`

I'm not sure what the alternative is here, but maybe there is a way to address these two (relatively minor) issues?

Reply:
I see your point. In this case secondary constructor route is the way to go